RFC/rfcBBL209/README.md

```diff
@@ -0,0 +1,104 @@
+
+* Status: `Brainstorm`
+
+## Abstract
+
+This RFC proposes adding a new type of data exchange to Bitswap for handling blocks of data arbitrarily larger than the 1MiB limit by using the features of common hash functions that allow for pausing and then resuming the hashes of large objects.
+
+## Shortcomings
+
+Bitswap has a maximum block size of 1MiB which means that it cannot transfer all forms of content addressed data. A prominent example of this is Git repos which even though they can be represented as a content addressed IPLD graph cannot necessarily be transferred over Bitswap if any of the objects in the repo exceed 1MiB.
+
+## Description
+
+The major hash functions work by taking some data `D` chunking it up into `n` pieces `P_0...P_n-1` then they modify an internal state `S` by loading pieces into the hash function in some way. This means that there are points in the hash function where we can pause processing and get the state of the hash function so far. Bitswap can utilize this state to effectively break up large blocks into smaller ones.
+
+### Example: Merkle–Damgård constructions like SHA-1 or SHA-2
+
+MD pseudo-code looks roughly like:
+
+```golang
+func Hash(D []byte) []byte {
+   pieces = getChunks(D)
+
+   var S state
+   for i, p := range pieces {
+      S = process(S, p) // Call this result S_i
+   }
+
+   return finalize(S) // Call this H, the final hash
+}
+```
+
+From the above we can see that:
+
+1. At any point in the process of hashing D we could stop, say after piece `j`, save the state `S_j` and then resume later
+2. We can always calculate the final hash `H` given only `S_j` and all the pieces `P_j+1..P_n-1`
+
+The implication for Bitswap is that if each piece size is not more than 1MiB then we can send the file **backwards** in 1MiB increments. In particular a server can send `(S_n-2, P_n-1)` and the client can use that to compute that `P_n-1` is in fact the last part of the data associated with the final hash `H`. The server can then send `(S_n-3, P_n-2)` and the client can calculate that `P_n-2` is the last block of `S_n-2` and therefore also the second to last block of `H`, and so on.
+
+#### Extension
+
+This scheme requires linearly downloading a file which is quite slow with even modest latencies. However, utilizing a scheme like [RFC|BB|L2 - Speed up Bitswap with GraphSync CID only query](https://github.com/protocol/beyond-bitswap/issues/25) (i.e. downloading metadata manifests up front) we can make this fast/parallelizable
+
+#### Security
+
+In order for this scheme to be secure it must be true that there is only a single pair `(S_i-1, P_i)` that can be produced to match with `S_i`. If the pair must be of the form `(S_i-1, P_malicious)` then this is certainly true since otherwise one could create a collision on the overall hash function. However, given that there are two parameters to vary it seems possible this could be computationally easier than finding a collision on the overall hash function.
```

Comment:
I'd like some 👀 on this ideally from people who are more practiced with this type of cryptanalysis than I am.

Reply:
I'm pretty sure that finding any kind of `(S_i-1, P_i)` pair that matches another `(S_i-1, P_other)` is at least as hard to crack as finding any arbitrary matching `(S_foo, P_foo)` `(S_bar, P_bar)` pair, which is the same as finding a hash collision between any two blocks. It is a birthday-attack, but secure hash functions have enough bits to be safe from birthday attacks.

Reply:
After some poking around (thanks @stebalien for pointers) it seems that as long as the underlying compression function is not subject to [freestart collisions](https://crypto.stackexchange.com/questions/29695/what-is-a-freestart-collision) then we should be fine and if not then things become trickier.

My understanding from this [paper](https://www.researchgate.net/publication/292963461_Analysis_of_SHA-512224_and_SHA-512256) and its corresponding [thesis](https://pure.tugraz.at/ws/portalfiles/portal/19191305/thesis_print_bib.pdf) is that SHA-256 is not yet subject to freestart collisions.

---

Even if we were subject to freestart collisions things may not necessarily be so bad since the attacker would also need to be the creator of the file and would be able to selectively give some people the data and other people would not get different data, but instead just waste some bandwidth and other resources which on its face doesn't seem like a super worthwhile attack.

If so then what we're really trying to avoid here is approximately a pseudo-second-preimage attack on the compressor function (close to the Definition 7 [here](https://www.iacr.org/archive/fse2012/75490268/75490268.pdf)). My understanding is that this would be even harder for an attacker to pull off and might even be reasonably safe for functions like SHA-1 which are no longer collision resistant (although pseudo-preimage attacks are of course may be easier to pull off than full preimage attacks).

---

@dbaarda thanks for the feedback, it does seem like this is probably ok. However, I do think it's a little more subtle than there are no collisions on SHA-2 implying there are no issues in this scheme. 

> `(S_foo, P_foo) (S_bar, P_bar)` pair, which is the same as finding a hash collision between any two blocks

My understanding is that this indicates a collision on the compressor function, but not on the overall hash function since a hash collision is that given some starting state IV `H(IV, P_good) = H(IV, P_bad)` (Definition 1 in that paper linked above).

Reply:
Ah, I did not know "freestart collisions" had a name; I'm glad it does!